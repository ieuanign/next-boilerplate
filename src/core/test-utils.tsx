import { GetServerSidePropsContext } from "next";
import { Router } from "next/router";
import { render } from "@testing-library/react";
import userEvent from "@testing-library/user-event";

import App from "@pages/_app.page";

const useRouter = jest.spyOn(require("next/router"), "useRouter");

jest.mock("react-i18next", () => ({
	useTranslation: () => ({ t: (key: any) => key }),
}));

export const setup = (Component: Function, defaultRouter?: Partial<Router>) => {
	let mockRouter = {
		pathname: "/",
		query: {},
		...defaultRouter,
		push: jest.fn(),
		prefetch: jest.fn(() => Promise.resolve(true)),
	};

	const withProps = (props?: any) => {
		const {
			hostname = "http://localhost:3000",
			pageProps = {},
			router,
			...rest
		} = props || {};

		mockRouter = { ...mockRouter, ...router };

		useRouter.mockReturnValue(mockRouter);

		const Comp = render(
			<App
				pageProps={pageProps}
				hostname={hostname}
				router={mockRouter}
				{...rest}
				Component={Component}
			/>
		);

		return {
			router: mockRouter,
			userEvent: userEvent.setup(),
			...Comp,
		};
	};

	return withProps;
};

export const getContext = (
	pathname: string,
	config?: any
): GetServerSidePropsContext =>
	({
		query: {},
		locale: "en",
		...config,
		resolvedUrl: pathname,
		req: {
			...config?.req,
			cookies: {
				...config?.req?.cookies,
			},
		},
	} as GetServerSidePropsContext);
