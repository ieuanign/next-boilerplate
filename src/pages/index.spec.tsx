import { screen, waitFor } from "@testing-library/react";
import { rest } from "msw";

import Home, { getServerSideProps } from "./index.page";

import { init, getContext, ssrLangTest } from "@core/test-utils";

const pathname = "/en";

const setup = init(Home, {
	pathname,
});

describe("getServerSideProps", () => {
	const defaultContext = getContext(pathname);

	ssrLangTest(defaultContext, getServerSideProps);
});

describe("Home", () => {
	it("should render with tasks link", async () => {
		setup();

		await waitFor(() => {
			expect(screen.getByText(/tasks/i)).toBeInTheDocument();
		});
	});
});
