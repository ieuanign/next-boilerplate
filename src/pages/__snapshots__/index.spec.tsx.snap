// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`getServerSideProps without locale 1`] = `
{
  "props": {
    "_nextI18Next": {
      "initialI18nStore": {
        "en": {
          "common": {
            "description": "NextJS common setup",
            "no": "No",
            "title": "NextJS Boilerplate | EN",
            "yes": "Yes",
          },
        },
      },
      "initialLocale": "en",
      "ns": [
        "common",
      ],
      "userConfig": {
        "default": {
          "i18n": {
            "defaultLocale": "default",
            "localeDetection": false,
            "locales": [
              "default",
              "en",
              "id",
            ],
          },
        },
        "i18n": {
          "defaultLocale": "default",
          "localeDetection": false,
          "locales": [
            "default",
            "en",
            "id",
          ],
        },
      },
    },
  },
}
`;
